astropy/cosmology/scalar_inv_efuncs.pyx

```diff
@@ -0,0 +1,222 @@
+    double wp, double apiv, double wa):
+  cdef double opz = 1.0 + z
+  cdef Odescl = opz**(3. * (1. + wp + apiv * wa)) * exp(-3. * wa * z / opz)
+  return (opz**2 * (opz * Om0 + Ok0) + Ode0 * Odescl)**(-0.5)
+
+# Massless neutrinos
+def wpwacdm_inv_efunc_nomnu(double z, double Om0, double Ode0, double Ok0,
+    double Or0, double wp, double apiv, double wa):
+  cdef double opz = 1.0 + z
+  cdef Odescl = opz**(3. * (1. + wp + apiv * wa)) * exp(-3. * wa * z / opz)
+  return ((((opz * Or0 + Om0) * opz) + Ok0) * opz**2 +
+          Ode0 * Odescl)**(-0.5)
+
+# With massive neutrinos
+def wpwacdm_inv_efunc(double z, double Om0, double Ode0, double Ok0,
+    double Ogamma0, double NeffPerNu, int nmasslessnu, list nu_y, double wp,
+    double apiv, double wa):
+
+  cdef double opz = 1.0 + z
+  cdef double Or0 = Ogamma0 * (1.0 + nufunc(opz, NeffPerNu, nmasslessnu, nu_y))
+  cdef Odescl = opz**(3. * (1. + wp + apiv * wa)) * exp(-3. * wa * z / opz)
+  return ((((opz * Or0 + Om0) * opz) + Ok0) * opz**2 + Ode0 * Odescl)**(-0.5)
+
+######## w0wzCDM
+# No relativistic species
+def w0wzcdm_inv_efunc_norel(double z, double Om0, double Ode0, double Ok0,
+    double w0, double wz):
+  cdef double opz = 1.0 + z
+  cdef Odescl = opz**(3. * (1. + w0 - wz)) * exp(-3. * wz * z)
+  return (opz**2 * (opz * Om0 + Ok0) + Ode0 * Odescl)**(-0.5)
+
+# Massless neutrinos
+def w0wzcdm_inv_efunc_nomnu(double z, double Om0, double Ode0, double Ok0,
+    double Or0, double w0, double wz):
+  cdef double opz = 1.0 + z
+  cdef Odescl = opz**(3. * (1. + w0 - wz)) * exp(-3. * wz * z)
+  return ((((opz * Or0 + Om0) * opz) + Ok0) * opz**2 +
+          Ode0 * Odescl)**(-0.5)
+
+# With massive neutrinos
+def w0wzcdm_inv_efunc(double z, double Om0, double Ode0, double Ok0,
+    double Ogamma0, double NeffPerNu, int nmasslessnu, list nu_y, double w0,
+    double wz):
+
+  cdef double opz = 1.0 + z
+  cdef double Or0 = Ogamma0 * (1.0 + nufunc(opz, NeffPerNu, nmasslessnu, nu_y))
+  cdef Odescl = opz**(3. * (1. + w0 - wz)) * exp(-3. * wz * z)
+  return ((((opz * Or0 + Om0) * opz) + Ok0) * opz**2 + Ode0 * Odescl)**(-0.5)
+
+######## Neutrino relative density function
+# Scalar equivalent to FLRW.nu_realative_density in core.py
+#  Please see that for further discussion.
+# This should only be called with massive neutrinos (e.g., nu_y is not empty)
+# Briefly, this is just a numerical fitting function to the true relationship,
+#  which is too expensive to want to evaluate directly.  The
+#  constants which appear are:
+#    p = 1.83  -> numerical fitting constant from Komatsu et al.
+#  1/p = 0.54644... -> same constant
+#    k = 0.3173 -> another fitting constant
+#  7/8 (4/11)^(4/3) = 0.2271... -> fermion/boson constant for neutrino
+#                                   contribution -- see any cosmology book
+#  The Komatsu reference is: Komatsu et al. 2011, ApJS 192, 18
+
+cdef nufunc(double opz, double NeffPerNu, int nmasslessnu, list nu_y):
```

Comment:
Well, even if that were good practice it wouldn't matter.  I played around with this a bit more and no matter what I did the array version was slower--amazingly.  Even with boundscheck and wraparound disabled, just the work involved in getting the array buffer and performing type checking ends up being slower, overall, for small arrays--for larger arrays you still get significant speedup using arrays.  One possibility would be to accept a generic object, and only if its size is large enough check if it's an array, and cast it as one.  But it's probably not worth the additional effort since this will almost always be small.

The unfortunate thing about how this function is defined is that since `nu_y` is a list, it has to treat each element of `nu_y` as a generic Python object, and so instead of calling C arithmetic operators it goes through the cumbersome Python operator handling.  However! I was able to get almost another 2x speedup simply by changing `nu_y[i]` to `<double>nu_y[i]`.  If Cython knows it can expect Python floats it can call PyFloat_AsDouble on them and do the rest of the arithmetic with pure C.